pkg/controller/logstorage/linseed/linseed_controller.go

```diff
@@ -0,0 +1,389 @@
+	"github.com/tigera/operator/pkg/render/logstorage/esgateway"
+	"github.com/tigera/operator/pkg/render/logstorage/linseed"
+	"github.com/tigera/operator/pkg/render/monitor"
+	"github.com/tigera/operator/pkg/tls/certificatemanagement"
+)
+
+var log = logf.Log.WithName("controller_logstorage_linseed")
+
+type LinseedSubController struct {
+	client         client.Client
+	scheme         *runtime.Scheme
+	status         status.StatusManager
+	clusterDomain  string
+	tierWatchReady *utils.ReadyFlag
+	usePSP         bool
+	multiTenant    bool
+}
+
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		return nil
+	}
+
+	// Create the reconciler
+	r := &LinseedSubController{
+		client:        mgr.GetClient(),
+		scheme:        mgr.GetScheme(),
+		clusterDomain: opts.ClusterDomain,
+		multiTenant:   opts.MultiTenant,
+		status:        status.New(mgr.GetClient(), "log-storage-access", opts.KubernetesVersion),
+	}
+	r.status.Run(opts.ShutdownContext)
+
+	// Create a controller using the reconciler and register it with the manager to receive reconcile calls.
+	c, err := controller.New("log-storage-access-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Determine how to handle watch events for cluster-scoped resources. For multi-tenant clusters,
+	// we should update all tenants whenever one changes. For single-tenatn clusters, we can just queue the object.
+	var eventHandler handler.EventHandler = &handler.EnqueueRequestForObject{}
+	if opts.MultiTenant {
+		eventHandler = utils.EnqueueAllTenants(mgr.GetClient())
+	}
+
+	// Configure watches for operator.tigera.io APIs this controller cares about.
+	if err = c.Watch(&source.Kind{Type: &operatorv1.LogStorage{}}, eventHandler); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch LogStorage resource: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.Installation{}}, eventHandler); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch Network resource: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementCluster{}}, eventHandler); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch ManagementCluster resource: %w", err)
+	}
+	if err = c.Watch(&source.Kind{Type: &operatorv1.ManagementClusterConnection{}}, eventHandler); err != nil {
+		return fmt.Errorf("log-storage-controller failed to watch ManagementClusterConnection resource: %w", err)
+	}
+	if err = utils.AddTigeraStatusWatch(c, "log-storage-access"); err != nil {
+		return fmt.Errorf("logstorage-controller failed to watch logstorage Tigerastatus: %w", err)
+	}
+	if opts.MultiTenant {
+		if err = c.Watch(&source.Kind{Type: &operatorv1.Tenant{}}, eventHandler); err != nil {
+			return fmt.Errorf("log-storage-controller failed to watch Tenant resource: %w", err)
+		}
+	}
+
+	// The namespace(s) we need to monitor depend upon what tenancy mode we're running in.
```

Comment:
This pattern is pretty ugly - I suspect there is a better way. Maybe with a helper interface that hides the namespace selection logic somehow?